src/main/java/org/truffleruby/core/string/TBytesKey.java

```diff
@@ -62,4 +87,51 @@ public String toString() {
         return TruffleString.fromByteArrayUncached(bytes, encoding, false).toString();
     }
 
+    private static int hashCode(InternalByteArray byteArray) {
+        return hashCode(byteArray.getArray(), byteArray.getOffset(), byteArray.getLength());
+    }
+
+    // A variant of <code>Arrays.hashCode</code> that allows for selecting a range within the array.
+    private static int hashCode(byte[] bytes, int offset, int length) {
+        if (bytes == null) {
+            return 0;
+        }
+
+        int result = 1;
+        for (int i = offset; i < offset + length; i++) {
+            result = 31 * result + bytes[i];
+        }
+
+        return result;
+    }
+
+    private boolean equalBytes(TBytesKey a, TBytesKey b) {
+        if (a.isPerfectFit() && b.isPerfectFit()) {
+            return Arrays.equals(a.bytes, b.bytes);
+        }
+
+        return Arrays.equals(a.bytes, a.offset, a.offset + a.length, b.bytes, b.offset, b.offset + b.length);
+    }
+
+    private boolean isPerfectFit() {
+        return offset == 0 && length == bytes.length;
+    }
+
+    public TBytesKey makeCacheable(boolean isImmutable) {
```

Comment:
Could you document this? (it creates a perfect fit key, so it does not hold on any extra bytes that it does not need when inserting a new entry to avoid leaking bytes outside the substring)